Feature fast cast-only mxfp8

Jianbing-D opened the pull request:
# Description

This pull request involves efficient implementations for mxfp8 quantize on `casting only` cases. It can increase the casting performance from 5%~ 20%.

It supports:
+ `BF16` or `FP16` as inputs
+ `E5M2` or `E4M3` as outputs
+  gpu arch >= `sm_100`
+ `rowwise` or `row- & col-wise`

Performance gain:
<img width="1670" height="328" alt="image" src="https://github.com/user-attachments/assets/b41f3ba0-76e7-475d-9a54-8b82a4c39391" />
<img width="1672" height="330" alt="image" src="https://github.com/user-attachments/assets/3d7f2d5d-860a-4247-ab6d-189eaeeace81" />
<img width="755" height="465" alt="image" src="https://github.com/user-attachments/assets/5f552f3e-b069-43e9-a306-52eaff3420b4" />
<img width="751" height="464" alt="image" src="https://github.com/user-attachments/assets/043fb0e5-df33-47e3-9b0c-4aa468e65a54" />


## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Added an environment `ENABLE_CAST_ONLY` to select optimized kernel. If optimized kernel doesn't support provided inputs, it will fallback to original kernels, automatically.
    1. If `ENABLE_CAST_ONLY ` is not set or is set to `0`, then original kernels will be used.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

ptrendx commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,2747 @@
+static_assert(sizeof(fp8e4m3x4) == 4);
+static_assert(sizeof(fp8e5m2x4) == 4);
+
+__device__ __forceinline__ void mul_cvt_4x(fp8e4m3x4 &out, const bf16x4 &in,
```
Why not have those functions in the ptx.cuh?

Jianbing-D replied:
> Why not have those functions in the ptx.cuh?

These functions are specialized for current file. It may not be reused elsewhere?

But if you think ptx.cuh is a better location, I can move them there.

ptrendx commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,2747 @@
+            }
+          }
+        }
+        // else {
```
Please remove commented out code.

Oleg-Goncharov reviewed (commented):
Solid kernel optimizations, the perf comes right when we need it.

Jianbing-D commented on transformer_engine/common/util/cast_kernels.cuh:
```diff
@@ -1081,6 +1082,72 @@ void mxfp8_quantize(const Tensor &input, const Tensor *act_input,
+
+                break;
+              }
+              case ScalingType::COLWISE: {
```
Replaced `NVTE_ERROR` with `NVTE_WARN("Colwise scaling will fallback to original kernel.");` when this `cast only opt is enabled`.

Jianbing-D commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,1685 @@
+
+namespace {
+#if (defined __CUDA_ARCH__) && (__CUDA_ARCH__ >= 1000)
+template <typename IType, typename OType>
```
No. The original `Quantized_limits` contains a `constexpr float` member called `max_norm_rcp`, But the `fma.rn.f32.f16` requires a `FP16` as input, which means the `constexpr float` should be converted to `constexpr fp16`. (If it is not `constexpr FP16`, there will be a unnecessary `cvt` instruction to do such conversion at runtime).

If do compile-time conversion in the following way, there will be a compile error:
```shell
../util/cast_kernels_spec.cuh(60): note #2703-D: cannot call non-constexpr function "__nv_bfloat16::__nv_bfloat16(float)" (declared at line 4505 of /usr/local/cuda/targets/sbsa-linux/include/cuda_bf16.h)
        constexpr auto max_norm_rcp = static_cast<IType>(Quantized_Limits<OType>::max_norm_rcp);
```

Jianbing-D commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,1685 @@
+              float2 amaxs;
+#if ((__CUDA_ARCH_HAS_FEATURE__(SM100_ALL)) || (__CUDA_ARCH_HAS_FEATURE__(SM101_ALL)) || \
+     (__CUDA_ARCH_HAS_FEATURE__(SM120_ALL)))
+              asm volatile("redux.sync.max.abs.f32 %0, %1, 0xFFFFFFFF;"
```
done

Jianbing-D commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,1685 @@
+                           : "=f"(amaxs.y)
+                           : "f"(values.y));
+#else
+              asm volatile(
```
done

Jianbing-D commented on transformer_engine/common/util/cast_kernels_spec.cuh:
```diff
@@ -0,0 +1,1685 @@
+      coords.x = block_coords.x + iter_n * CastTraits::blockIterDimN;
+
+      if (coords.y < rows && coords.x < cols) {
+        int32_t offset = coords.y * cols + coords.x;
```
replaced `int32_t` with `size_t` for global memory pointer offsetting.

Oleg-Goncharov commented:
/te-ci

Jianbing-D commented:
> /te-ci

Hi @Oleg-Goncharov , looks like there were issues on the CI system.
```log
Installing collected packages: nvidia-mathdx, nvidia-cusparselt-cu12, mpmath, triton, sympy, pybind11-global, pybind11, nvidia-nvtx-cu12, nvidia-nvshmem-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufile-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, networkx, MarkupSafe, filelock, einops, onnx, nvidia-cusparse-cu12, nvidia-cufft-cu12, nvidia-cudnn-cu12, jinja2, onnx_ir, nvidia-cusolver-cu12, torch, onnxscript
ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device
```

Oleg-Goncharov commented:
Hi @Jianbing-D, could you please resolve the merge conflicts once more? Sorry for asking to do this again, but we refactored the cast kernels to organize them into clearer groups, so your MXFP8 kernel should now live under `/common/cast/mxfp8/specialized/` with the appropriate kernel filtering. CI is fixed, so once the conflicts are resolved I can merge. Thanks a lot!

Jianbing-D commented:
> Hi @Jianbing-D, could you please resolve the merge conflicts once more? Sorry for asking to do this again, but we refactored the cast kernels to organize them into clearer groups, so your MXFP8 kernel should now live under `/common/cast/mxfp8/specialized/` with the appropriate kernel filtering. CI is fixed, so once the conflicts are resolved I can merge. Thanks a lot!

Hi @Oleg-Goncharov , I have refactored this branch, and put my kernel inside the folder you suggested. Please review it. Thanks.

Oleg-Goncharov commented:
/te-ci
